src/Electrical/Digital/tables.jl

```diff
@@ -1,131 +1,97 @@
-    OrTable[i, j]
+_not(x::Logic) = NotTable[get_logic_level(x)]
+_not(x::Number) = _not(convert_to_logic(x))
+
+@register_symbolic _not(x)
+
+# OR gate
+OrTable = [
+           # U X F0 F1 Z W L H DC
+           U U U F1 U U U F1 U        # U
+           U X X F1 X X X F1 X        # X
+           U X F0 F1 X X F0 F1 X      # F0
+           F1 F1 F1 F1 F1 F1 F1 F1 F1 # F1
+           U X X F1 X X X F1 X        # Z
+           U X X F1 X X X F1 X        # W
+           U X F0 F1 X X F0 F1 X      # L
+           F1 F1 F1 F1 F1 F1 F1 F1 F1 # H
+           U X X F1 X X X F1 X]       # DC
+
+function _or2(a::Logic, b::Logic)
+    OrTable[get_logic_level(a), get_logic_level(b)]
 end
+_or2(a::Number, b::Logic) = _or2(convert_to_logic(a), b)
+_or2(a::Logic, b::Number) = _or2(a, convert_to_logic(b))
+_or2(a::Number, b::Number) = _or2(convert_to_logic(a), convert_to_logic(b))
 
 function _or(x...)
-    y = [_or2(x[1], x[1])]
-    for i in 2:length(x)
+    y = [convert_to_logic(x[1])]
+    for i in 2:lastindex(x)
         push!(y, _or2(x[i], y[i - 1]))
     end
     return y[end]
 end
-@register_symbolic _or(x...)
-@register_symbolic _or(a, b)
-@register_symbolic _or(a, b, c)
-@register_symbolic _or(a, b, c, d)
-@register_symbolic _or(a, b, c, d, e)
-@register_symbolic _or(a, b, c, d, e, f, g, h)
 
-# MISO :XOR gate
+@register_symbolic _or(a, b)
 
-XorTable = OffsetArray([
-                        #  0  1 :U :X :Z :W :L :H :D
-                        0 1 :U :X :X :X 0 1 :X # 0
-                        1 0 :U :X :X :X 1 0 :X # 1
-                        :U :U :U :U :U :U :U :U :U # :U
-                        :X :X :U :X :X :X :X :X :X # :X
-                        :X :X :U :X :X :X :X :X :X # :Z
-                        :X :X :U :X :X :X :X :X :X # :W
-                        0 1 :U :X :X :X 0 1 :X # :L
-                        1 0 :U :X :X :X 1 0 :X # :H
-                        :X :X :U :X :X :X :X :X :X], 0:8, 0:8)
+# XOR gate
+XorTable = [
+            # U X F0 F1 Z W L H DC
+            U U U U U U U U U     # U
+            U X X X X X X X X     # X
+            U X F0 F1 X X F0 F1 X # F0
+            U X F1 F0 X X F1 F0 X # F1
+            U X X X X X X X X     # Z
+            U X X X X X X X X     # W
+            U X F0 F1 X X F0 F1 X # L
+            U X F1 F0 X X F1 F0 X # H
+            U X X X X X X X X]    # DC
 
-function _xor2(a, b)
-    i, j = logic(a), logic(b)
-    (typeof(i) != Int || typeof(j) != Int) && return X
-    XorTable[i, j]
+function _xor2(a::Logic, b::Logic)
+    XorTable[get_logic_level(a), get_logic_level(b)]
 end
+_xor2(a::Number, b::Logic) = _xor2(convert_to_logic(a), b)
+_xor2(a::Logic, b::Number) = _xor2(a, convert_to_logic(b))
+_xor2(a::Number, b::Number) = _xor2(convert_to_logic(a), convert_to_logic(b))
 
 function _xor(x...)
-    y = [_xor2(x[1], 0)]
-    for i in 2:length(x)
+    y = [convert_to_logic(x[1])]
+    for i in 2:lastindex(x)
         push!(y, _xor2(x[i], y[i - 1]))
     end
     return y[end]
```

Comment:
You're only ever using the last value of `y`, so do you really need to allocate an array?

Reply:
Good point